### PR TITLE
Clean up Xcode data types

### PR DESCRIPTION
**Pull request by mdiep**
Starting to clean things up a bit.

### CODE REVIEW

File: Source/CarthageKitTests/XcodeSpec.swift [hunk 1]
```diff
@@ -58,6 +58,32 @@ class XcodeSpec: QuickSpec {
 			_ = try? NSFileManager.defaultManager().removeItemAtURL(targetFolderURL)
 			return
 		}
+		
+		describe("\(ProjectLocator.self)") {
+			describe("sorting") {
+				it("should put workspaces before projects") {
+					let workspace = ProjectLocator.Workspace(NSURL(fileURLWithPath: "/Z.xcworkspace"))
+					let project = ProjectLocator.ProjectFile(NSURL(fileURLWithPath: "/A.xcodeproj"))
+					expect(workspace < project).to(beTrue())
```
Comment (ikesyo) [inline Source/CarthageKitTests/XcodeSpec.swift]: We are using `beTruthy()` in the existing specs.

File: Source/CarthageKitTests/XcodeSpec.swift [hunk 2]
```diff
@@ -58,6 +58,32 @@ class XcodeSpec: QuickSpec {
...
+				}
+				
+				it("should fall back to lexicographical sorting") {
+					let workspaceA = ProjectLocator.Workspace(NSURL(fileURLWithPath: "/A.xcworkspace"))
+					let workspaceB = ProjectLocator.Workspace(NSURL(fileURLWithPath: "/B.xcworkspace"))
+					expect(workspaceA < workspaceB).to(beTrue())
```
Comment (ikesyo) [inline Source/CarthageKitTests/XcodeSpec.swift]: Ditto.

File: Source/CarthageKitTests/XcodeSpec.swift [hunk 3]
```diff
@@ -58,6 +58,32 @@ class XcodeSpec: QuickSpec {
...
+					
+					let projectA = ProjectLocator.ProjectFile(NSURL(fileURLWithPath: "/A.xcodeproj"))
+					let projectB = ProjectLocator.ProjectFile(NSURL(fileURLWithPath: "/B.xcodeproj"))
+					expect(projectA < projectB).to(beTrue())
```
Comment (ikesyo) [inline Source/CarthageKitTests/XcodeSpec.swift]: Ditto.

File: Source/CarthageKitTests/XcodeSpec.swift [hunk 4]
```diff
@@ -58,6 +58,32 @@ class XcodeSpec: QuickSpec {
...
+				}
+				
+				it("should put top-level directories first") {
+					let top = ProjectLocator.ProjectFile(NSURL(fileURLWithPath: "/Z.xcodeproj"))
+					let bottom = ProjectLocator.Workspace(NSURL(fileURLWithPath: "/A/A.xcodeproj"))
+					expect(top < bottom).to(beTrue())
```
Comment (ikesyo) [inline Source/CarthageKitTests/XcodeSpec.swift]: Ditto.

File: Source/CarthageKit/Xcode.swift [hunk 5]
```diff
@@ -199,21 +183,23 @@ public func locateProjectsInDirectory(directoryURL: NSURL) -> SignalProducer<Pro
...
 
 	return NSFileManager.defaultManager().carthage_enumeratorAtURL(directoryURL.URLByResolvingSymlinksInPath!, includingPropertiesForKeys: [ NSURLTypeIdentifierKey ], options: enumerationOptions, catchErrors: true)
-		.reduce([]) { (var matches: [ProjectEnumerationMatch], tuple) -> [ProjectEnumerationMatch] in
-			let (enumerator, URL) = tuple
-			if let match = ProjectEnumerationMatch.matchURL(URL, fromEnumerator: enumerator).value {
-				matches.append(match)
+		.reduce([]) { (var matches: [ProjectLocator], tuple) -> [ProjectLocator] in
+			let (_, URL) = tuple
+			
+			if let UTI = URL.typeIdentifier.value {
```
Comment (ikesyo) [inline Source/CarthageKit/Xcode.swift]: Is not `URL.URLByResolvingSymlinksInPath` needed?

File: Source/CarthageKit/Xcode.swift [hunk 6]
```diff
@@ -154,44 +176,6 @@ extension BuildArguments: CustomStringConvertible {
...
-	static func matchURL(URL: NSURL, fromEnumerator enumerator: NSDirectoryEnumerator) -> Result<ProjectEnumerationMatch, CarthageError> {
-		if let URL = URL.URLByResolvingSymlinksInPath {
-			return URL.typeIdentifier.flatMap { typeIdentifier in
-				if (UTTypeConformsTo(typeIdentifier, "com.apple.dt.document.workspace")) {
-					return .Success(ProjectEnumerationMatch(locator: .Workspace(URL), level: enumerator.level))
-				} else if (UTTypeConformsTo(typeIdentifier, "com.apple.xcode.project")) {
-					return .Success(ProjectEnumerationMatch(locator: .ProjectFile(URL), level: enumerator.level))
-				}
-
-				return .Failure(.NotAProject(URL))
```
Comment (ikesyo) [inline Source/CarthageKit/Xcode.swift]: `CarthageError.NotAProject` isn't used anymore, so we can remove the case.

File: Source/CarthageKit/Xcode.swift [hunk 7]
```diff
@@ -64,6 +52,15 @@ public func ==(lhs: ProjectLocator, rhs: ProjectLocator) -> Bool {
 }
 
 public func <(lhs: ProjectLocator, rhs: ProjectLocator) -> Bool {
+	// Prefer top-level directories
+	let leftLevel = lhs.fileURL.pathComponents!.count
```
Comment (ikesyo) [inline Source/CarthageKit/Xcode.swift]: How about adding `public var level: Int` to `ProjectLocator`?

File: Source/CarthageKit/Xcode.swift [hunk 8]
```diff
@@ -575,12 +500,14 @@ public struct BuildSettings {
 							currentTarget = (line as NSString).substringWithRange(targetRange)
 							return
 						}
-
-						let components = line.characters.split(1) { $0 == "=" }.map(String.init)
+						
 						let trimSet = NSCharacterSet.whitespaceAndNewlineCharacterSet()
+						let components = line.characters
+							.split(1) { $0 == "=" }
+							.map { String($0).stringByTrimmingCharactersInSet(trimSet) }
```
Comment (ikesyo) [inline Source/CarthageKit/Xcode.swift]: It's cool. :100:

**Comment by ikesyo** [comment]
Thank you for the effor! :sparkles: Just a few minor comments.

File: Source/CarthageKit/Xcode.swift (hunk 6 above)
Comment (mdiep) [inline Source/CarthageKit/Xcode.swift]: :hocho:'d!

File: Source/CarthageKit/Xcode.swift (hunk 5 above)
Comment (mdiep) [inline Source/CarthageKit/Xcode.swift]: I can't say for sure. No tests are failing though. :trollface: 

But I'd rather simplify the code until and unless we find a failing case. Resolving symlinks is arguably a job for the consumer of `ProjectLocator`s anyway.

File: Source/CarthageKit/Xcode.swift (hunk 7 above)
Comment (mdiep) [inline Source/CarthageKit/Xcode.swift]: :+1:

**Comment by mdiep** [comment]
🌮

File: Source/CarthageKit/Xcode.swift (hunk 5 above)
Comment (ikesyo) [inline Source/CarthageKit/Xcode.swift]: I'm sure. :ok_hand:

**Comment by ikesyo** [comment]
:shipit:

**Pull request merged**